gobblin-service/src/main/java/org/apache/gobblin/service/monitoring/KafkaJobStatusMonitor.java

```diff
@@ -253,14 +253,22 @@ static void addJobStatusToStateStore(org.apache.gobblin.configuration.State jobS
 
     List<org.apache.gobblin.configuration.State> states = stateStore.getAll(storeName, tableName);
     if (states.size() > 0) {
-      String previousStatus = states.get(states.size() - 1).getProp(JobStatusRetriever.EVENT_NAME_FIELD);
+      org.apache.gobblin.configuration.State previousJobStatus = states.get(states.size() - 1);
+      String previousStatus = previousJobStatus.getProp(JobStatusRetriever.EVENT_NAME_FIELD);
       String currentStatus = jobStatus.getProp(JobStatusRetriever.EVENT_NAME_FIELD);
-
-      // PENDING_RESUME is allowed to override, because it happens when a flow is being resumed from previously being failed
-      if (previousStatus != null && currentStatus != null && !currentStatus.equals(ExecutionStatus.PENDING_RESUME.name())
-        && ORDERED_EXECUTION_STATUSES.indexOf(ExecutionStatus.valueOf(currentStatus)) < ORDERED_EXECUTION_STATUSES.indexOf(ExecutionStatus.valueOf(previousStatus))) {
-        log.warn(String.format("Received status %s when status is already %s for flow (%s, %s, %s), job (%s, %s)",
-            currentStatus, previousStatus, flowGroup, flowName, flowExecutionId, jobGroup, jobName));
+      int previousGeneration = previousJobStatus.getPropAsInt(TimingEvent.FlowEventConstants.CURRENT_GENERATION_FIELD, 1);
+      int currentGeneration = jobStatus.getPropAsInt(TimingEvent.FlowEventConstants.CURRENT_GENERATION_FIELD, 1);
+      int previousAttempts = previousJobStatus.getPropAsInt(TimingEvent.FlowEventConstants.CURRENT_ATTEMPTS_FIELD, 1);
+      int currentAttempts = jobStatus.getPropAsInt(TimingEvent.FlowEventConstants.CURRENT_ATTEMPTS_FIELD, 1);
+      // We use three things to determine the order, generation means how many times job has been resumed, attempts means during this generation, how many times job has been retried
```

Comment:
minor, but it seems not so much "to determine the order", but "to accurately count and thereby bound retries, even amidst out-of-order events (by skipping late arrivals). generation tracks..."

then later explain, "the generation is monotonically increasing, while the attempts may re-initialize back to 0.  this two-part form prevents the composite value from ever repeating."  (or something similar...)